src/utils/debounce.ts

```diff
@@ -0,0 +1,146 @@
+interface DebounceOptions {
+  maxWait?: number;
+  leading?: boolean;
+  trailing?: boolean;
+}
+
+export interface IDebounceReturn {
+  cancel: () => void;
+  flush: () => void;
+  isPending: () => boolean;
+}
+
+export interface DebouncedReturn<T extends (...args: any[]) => ReturnType<T>> extends IDebounceReturn {
+  (...args: Parameters<T>): ReturnType<T> | undefined;
+}
+
+export function debounce<T extends (...args: any) => ReturnType<T>>(fn: T, wait?: number, options?: DebounceOptions): DebouncedReturn<T> {
```

Comment:
I don't think we need an implementation that covers every corner case. We just need a minimum implementation like:

```js
// we only used the `leading` option in our code
// NEED TEST!
function debounce(fn, wait, leading) {
  let timer;

  return (...args) => {
    clearTimeout(timer);
 
    let ms = leading ? 0 : wait;
    leading = false;

    timer = setTimeout(() => {
      fn(...args);
    }, ms);
  };
}
```

Reply:
See also: https://css-tricks.com/debouncing-throttling-explained-examples/

Reply:
lodash **`useRaf`** pushed me to copy same logic, cause most of `debounce` utils on web using similar variables and logics as lodash debounce

what is ur idea about refactor same code you shared with both `setTimeout` and `requestAnimationFrame`?

cause `requestAnimationFrame` is more efficient than `setTimeout`

Reply:
GitHub Copilot debounce suggestion :alien:
```js
function debounce(func, wait, leading) {
  let timeout
  return function () {
    const context = this; const args = arguments
    const later = function () {
      timeout = null
      if (!leading)
        func.apply(context, args)
    }
    const callNow = leading && !timeout
    clearTimeout(timeout)
    timeout = setTimeout(later, wait)
    if (callNow)
      func.apply(context, args)
  }
}
```

Reply:
Copilot's version is functionally the same as mine except that I forgot to support `this` context. 😆

Reply:
Fixed version:

```js
// we only used the `leading` option in our code
// NEED MORE TESTS!
function debounce(fn, wait, leading) {
  let timer;
  let lastCalledAt = -Infinity;

  function handleLeading(thisArg, args) {
    const now = Date.now();
    const elapsed = now - lastCalledAt;

    lastCalledAt = now;

    if (elapsed >= wait) {
      fn.apply(thisArg, args);
    }
  }

  return function debouncedFn(...args) {
    if (leading) {
      handleLeading(this, args);
      return;
    }

    clearTimeout(timer);

    timer = setTimeout(() => {
      fn.apply(this, args);
    }, wait);
  };
}

```